server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -668,10 +669,26 @@ void validatePipeline(Map<DiscoveryNode, IngestInfo> ingestInfos, String pipelin
         ExceptionsHelper.rethrowAndSuppress(exceptions);
     }
 
+    private record IngestPipelinesExecutionResult(boolean success, boolean kept, Exception exception, String failedIndex) {}
+
+    private static IngestPipelinesExecutionResult successResult() {
+        return new IngestPipelinesExecutionResult(true, true, null, null);
+    }
+
+    private static IngestPipelinesExecutionResult discardResult() {
+        return new IngestPipelinesExecutionResult(true, false, null, null);
+    }
+
+    private static IngestPipelinesExecutionResult failAndStoreFor(String index, Exception e) {
+        return new IngestPipelinesExecutionResult(false, true, e, index);
+    }
+
     public void executeBulkRequest(
```

Comment:
I think this method is now complex enough that we need javadoc to explain the difference between the arguments